### PR TITLE
Feat/issue 21 add app param api

### PR DESCRIPTION
**Pull request by abbasogaji**
# Issue

 [ #21 ](https://github.com/asos-craigmorten/opine/issues/21)

## Details

Brief summary of PR purpose and code changes.

added 
- app.param([name], callback) 

**"Add callback triggers to route parameters, where name is the name of the parameter or an array of them, and callback is the callback function. The parameters of the callback function are the request object, the response object, the next middleware, the value of the parameter and the name of the parameter, in that order."** [ source. [expressjs docs](https://expressjs.com/en/4x/api.html#app.param) ]

added new type definition for app.param()  at - src/types.ts
added cloned implementation of expressjs app.param() at - src/application.ts
added cloned  implementation of expressjs router app.param() at src/router/index.ts
added unit test for app.param() api at - test/units/app.param.test.ts  (cloned from expressjs)
 
## CheckList

- [x] PR starts with [#21 ].
- [x] Has been tested (where required) before merge to main.

### CODE REVIEW

**Comment by asos-craigmorten** [comment]
@abbasogaji looks like CI is failing with lint errors - `make fmt` should do the trick!

**Comment by abbasogaji** [comment]
> @abbasogaji looks like CI is failing with lint errors - `make fmt` should do the trick!

 - still failed after running `make fmt`. on ci when trying to format "/opine/docs/assets/js/main.js"
 - while everything seems to be fine locally

**Comment by abbasogaji** [comment]
> > @abbasogaji looks like CI is failing with lint errors - `make fmt` should do the trick!
> > @asos-craigmorten
> 
> * still failed via github actions, but ok locally.
> * while everything seems to be fine locally. when i run `make fmt` locally nothing new happens
>   [FIXED]

@asos-craigmorten Fixed:  issue occurred cause i was using an older deno version, so format error didnt show up on version 1.0.0

**Comment by asos-craigmorten** [comment]
Ah good catch, we’re currently supporting 1.2.0 which is somewhat different to the initial release version!

File: docs/_config.yaml [hunk 1]
```diff
@@ -1,5 +1 @@
-future: true
-encoding: "UTF-8"
-include:
-  - "_*_.html"
-  - "_*_.*.html"
+future: true\nencoding: "UTF-8"\ninclude:\n  - "_*_.html"\n  - "_*_.*.html"
```
Comment (asos-craigmorten) [inline docs/_config.yaml:1]: This doesn’t look right...

File: src/router/index.ts [hunk 2]
```diff
@@ -44,6 +44,73 @@ export const Router: RouterConstructor = function (options: any = {}): any {
 } as any;
 
 // TODO: Router.param
```
Comment (asos-craigmorten) [inline src/router/index.ts:46]:
```suggestion
```

This can go 😊

File: src/router/index.ts [hunk 3]
```diff
@@ -44,6 +44,73 @@ export const Router: RouterConstructor = function (options: any = {}): any {
...
+    return;
+  }
+
+  // apply param functions
+  var params = this._params;
+  var len = params.length;
+  var ret;
+
+  if (name[0] === ":") {
+    // deprecate('router.param(' + JSON.stringify(name) + ', fn): Use router.param(' + JSON.stringify(name.substr(1)) + ', fn) instead');
```
Comment (asos-craigmorten) [inline src/router/index.ts:95]: We should avoid supporting an paths in Opine that have been deprecated in Express and likely will be removed a next version. In this instance, we can drop support for `app.param(fn)` signature above and this support for `:` prefixed names.

Comment (abbasogaji) [reply src/router/index.ts:94]: Oh yeah right👍

File: docs/_config.yaml (hunk 1 above)
Comment (abbasogaji) [reply docs/_config.yaml:1]:
```makefile
typedoc:
	@rm -rf docs
	@typedoc --ignoreCompilerErrors --out ./docs --mode modules --includeDeclarations --excludeExternals --name opine ./src
	@make fmt
	@make fmt
	@echo 'future: true\nencoding: "UTF-8"\ninclude:\n  - "_*_.html"\n  - "_*_.*.html"' > ./docs/_config.yaml

```
hmmm i see, there is a line in the makefile where you write that line into doc/_config.yaml file, see above,
unless this works diferently on windows

Comment (abbasogaji) [reply docs/_config.yaml:1]: but i see what you trying to do, i am going to update that manually

**Comment by abbasogaji** [comment]
@asos-craigmorten

File: docs/_config.yaml (hunk 1 above)
Comment (asos-craigmorten) [reply docs/_config.yaml]: Ta - It’s clearly not having the desired effect on windows, works on MacOS / Linux. Unfortunately don’t have a windows PC to test with!

Comment (abbasogaji) [reply docs/_config.yaml]: @asos-craigmorten ok Fixed 🙂

File: src/router/index.ts (hunk 3 above)
Comment (abbasogaji) [reply src/router/index.ts]: @asos-craigmorten fixed 🙂

**Comment by asos-craigmorten** [comment]
LGTM - tests are failing due to unrelated breaking change with deno.land/x registry. Merging and will fix in a separate PR>

**Pull request merged**